File: dashboard/src/lib/mcp.ts
/**
 * MCP Client — typed wrapper around JSON-RPC calls to the alpha-swarm MCP server.
 *
 * All dashboard data flows through this single client.
 * Transport: HTTP POST to /mcp endpoint (MCP Streamable HTTP).
 */

const DEFAULT_MCP_URL = "/mcp";

// --- JSON-RPC types ---

interface JsonRpcRequest {
  jsonrpc: "2.0";
  id: number;
  method: string;
  params?: Record<string, unknown>;
}

interface JsonRpcResponse<T = unknown> {
  jsonrpc: "2.0";
  id: number;
  result?: T;
  error?: { code: number; message: string };
}

// --- MCP response types ---

interface McpToolResult {
  content: Array<{ type: string; text: string }>;
}

interface McpResourceResult {
  contents: Array<{ uri: string; mimeType: string; text: string }>;
}

// --- Client ---

let requestId = 0;
let mcpUrl = DEFAULT_MCP_URL;
let sessionId: string | null = null;

export function configure(url: string) {
  mcpUrl = url;
}

async function rpc<T>(method: string, params?: Record<string, unknown>): Promise<T> {
  const req: JsonRpcRequest = {
    jsonrpc: "2.0",
    id: ++requestId,
    method,
    params,
  };

  const resp = await fetch(mcpUrl, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      ...(sessionId ? { "mcp-session-id": sessionId } : {}),
    },
    body: JSON.stringify(req),
  });

  // Capture session ID from response
  const sid = resp.headers.get("mcp-session-id");
  if (sid) sessionId = sid;

  const json: JsonRpcResponse<T> = await resp.json();

  if (json.error) {
    throw new Error(`MCP error ${json.error.code}: ${json.error.message}`);
  }

  return json.result as T;
}

// --- Initialize ---

export async function initialize() {
  return rpc<{
    protocolVersion: string;
    serverInfo: { name: string; version: string };
    capabilities: Record<string, unknown>;
  }>("initialize", {
    protocolVersion: "2025-11-25",
    clientInfo: { name: "alpha-swarm-dashboard", version: "0.1.0" },
  });
}

// --- Tools (actions) ---

function extractText(result: McpToolResult): string {
  return result.content.map((c) => c.text).join("\n");
}

export async function submitTask(project: string, goal: string): Promise<string> {
  const result = await rpc<McpToolResult>("tools/call", {
    name: "submit_task",
    arguments: { project, goal },
  });
  return extractText(result);
}

export async function approveTask(runId: string): Promise<string> {
  const result = await rpc<McpToolResult>("tools/call", {
    name: "approve_task",
    arguments: { run_id: runId },
  });
  return extractText(result);
}

export async function deleteTask(runId: string): Promise<string> {
  const result = await rpc<McpToolResult>("tools/call", {
    name: "delete_task",
    arguments: { run_id: runId },
  });
  return extractText(result);
}

export async function planFeedback(runId: string, feedback: string): Promise<string> {
  const result = await rpc<McpToolResult>("tools/call", {
    name: "plan_feedback",
    arguments: { run_id: runId, feedback },
  });
  return extractText(result);
}

export async function createProject(name: string, repoUrl: string, description?: string): Promise<string> {
  const result = await rpc<McpToolResult>("tools/call", {
    name: "create_project",
    arguments: { name, repo_url: repoUrl, ...(description ? { description } : {}) },
  });
  return extractText(result);
}

export async function deleteProject(name: string): Promise<string> {
  const result = await rpc<McpToolResult>("tools/call", {
    name: "delete_project",
    arguments: { name },
  });
  return extractText(result);
}

// --- Resources (read-only data) ---

// SurrealDB REST returns [{result: [...], status: "OK"}]
// MCP resource passes this through as text. We unwrap it.
interface SurrealResponse<T> {
  result: T[];
  status: string;
}

function findLastArrayResult<T>(responses: SurrealResponse<T>[]): T[] {
  for (let i = responses.length - 1; i >= 0; i--) {
    if (Array.isArray(responses[i].result)) {
      return responses[i].result;
    }
  }
  return [];
}

function isSurrealFormat(parsed: unknown): boolean {
  return Array.isArray(parsed) && parsed.length > 0 && "result" in (parsed as Record<string, unknown>[])[0];
}

function parseResourceJson<T>(result: McpResourceResult): T[] {
  const text = result.contents[0]?.text ?? "[]";
  const parsed = JSON.parse(text);
  if (isSurrealFormat(parsed)) {
    return findLastArrayResult<T>(parsed as SurrealResponse<T>[]);
  }
  if (Array.isArray(parsed)) return parsed as T[];
  return [parsed as T];
}

export async function readResource<T = unknown>(uri: string): Promise<T[]> {
  const result = await rpc<McpResourceResult>("resources/read", { uri });
  return parseResourceJson<T>(result);
}

import type { Project, AgentRun, DashboardStats, GoalPlan } from "@/types/swarm";

// Typed resource helpers — all return unwrapped arrays
export const resources = {
  projects: () => readResource<Project>("swarm://projects"),
  models: () => readResource("swarm://models"),
  systemResources: () => readResource("swarm://resources"),
  health: () => readResource("swarm://health"),
  live: () => readResource<AgentRun>("swarm://live"),
  dashboard: () => readResource<DashboardStats>("swarm://dashboard"),

  projectRuns: (project: string) => readResource<AgentRun>(`swarm://projects/${project}/runs`),
  projectGoals: (project: string) => readResource(`swarm://projects/${project}/goals`),
  projectMetrics: (project: string) => readResource(`swarm://projects/${project}/metrics`),

  runDetail: (id: string) => readResource<AgentRun>(`swarm://runs/${id}`),
  subRuns: (id: string) => readResource<AgentRun>(`swarm://runs/${id}/sub-runs`),
  plans: (id: string) => readResource<GoalPlan>(`swarm://runs/${id}/plans`),
  timeline: (id: string) => readResource(`swarm://runs/${id}/timeline`),
};

File: components/mcp-server/src/mcp.rs
//! MCP JSON-RPC 2.0 protocol handler.
//!
//! Implements the MCP spec (2025-11-25) Streamable HTTP transport.
//! Methods: initialize, tools/list, tools/call, resources/list, resources/read.

use serde_json::{json, Value};

/// Protocol version we implement.
pub const MCP_PROTOCOL_VERSION: &str = "2025-11-25";
const SERVER_NAME: &str = "alpha-swarm";
const SERVER_VERSION: &str = "0.1.0";

/// Handle a JSON-RPC 2.0 request and return a JSON-RPC response.
pub fn handle_jsonrpc(body: &str) -> String {
    let req: Value = match serde_json::from_str(body) {
        Ok(v) => v,
        Err(e) => return jsonrpc_error(Value::Null, -32700, &format!("Parse error: {e}")),
    };

    let id = req.get("id").cloned().unwrap_or(Value::Null);
    let method = req.get("method").and_then(|m| m.as_str()).unwrap_or("");
    let params = req.get("params").cloned().unwrap_or(json!({}));

    let result = match method {
        "initialize" => handle_initialize(&params),
        "notifications/initialized" => return String::new(), // notification, no response
        "ping" => Ok(json!({})),
        "tools/list" => handle_tools_list(),
        "tools/call" => handle_tools_call(&params),
        "resources/list" => handle_resources_list(),
        "resources/read" => handle_resources_read(&params),
        "prompts/list" => Ok(json!({"prompts": []})),
        _ => Err(format!("Method not found: {method}")),
    };

    match result {
        Ok(result) => jsonrpc_ok(id, result),
        Err(e) => jsonrpc_error(id, -32601, &e),
    }
}

fn handle_initialize(_params: &Value) -> Result<Value, String> {
    Ok(json!({
        "protocolVersion": MCP_PROTOCOL_VERSION,
        "serverInfo": {
            "name": SERVER_NAME,
            "version": SERVER_VERSION,
        },
        "capabilities": {
            "tools": { "listChanged": false },
            "resources": { "subscribe": false, "listChanged": false },
        }
    }))
}

fn handle_tools_list() -> Result<Value, String> {
    Ok(json!({
        "tools": crate::tools::list_tools()
    }))
}

fn handle_tools_call(params: &Value) -> Result<Value, String> {
    let name = params.get("name").and_then(|n| n.as_str()).unwrap_or("");
    let args = params.get("arguments").cloned().unwrap_or(json!({}));
    crate::tools::call_tool(name, &args)
}

fn handle_resources_list() -> Result<Value, String> {
    Ok(json!({
        "resources": crate::resources::list_resources()
    }))
}

fn handle_resources_read(params: &Value) -> Result<Value, String> {
    let uri = params.get("uri").and_then(|u| u.as_str()).unwrap_or("");
    crate::resources::read_resource(uri)
}

// --- JSON-RPC helpers ---

fn jsonrpc_ok(id: Value, result: Value) -> String {
    json!({
        "jsonrpc": "2.0",
        "id": id,
        "result": result,
    }).to_string()
}

fn jsonrpc_error(id: Value, code: i32, message: &str) -> String {
    json!({
        "jsonrpc": "2.0",
        "id": id,
        "error": {
            "code": code,
            "message": message,
        }
    }).to_string()
}
